### DISCUSSION TITLE
I Want to Create a Voxel World Similar to Minecraft, But the FPS Is Very Low

### DISCUSSION THREAD

**Discussion by MrTwoc** [bevyengine/bevy#16215] [Q&A]
Just started learning programming, so please forgive my poor code.
Introduction: I intend to use the Bevy engine to create a world generated by noise, but currently, the FPS is very low, and I'm unsure of the reason.
The current generation scope should be 12832128, and the FPS is approximately 18~20 FPS.
I want to know what I need to learn or which demos I can refer to in order to improve the FPS.
Note: Apologies for my poor English; this is the translated text.
![36DC677D8895F8EB0060E16A1C958898](https://github.com/user-attachments/assets/4fe5da50-083e-4b2d-985d-ed856493e583)

Here is all the code from my project:

`use bevy::{
    dev_tools::fps_overlay::{FpsOverlayConfig, FpsOverlayPlugin}, prelude::*, render::render_resource::Face
};
use bevy_flycam::prelude::*;
use fastnoise_lite::*;

#[derive(Component)]
struct Block{
    block_id: u32,
    block_type: BlockType,
}
#[derive(Component)]
struct BlockType{
    block_name: String,
    x: u32,
    y: u32,
    z: u32,
}

fn main() {
    App::new()
        .add_plugins((
            DefaultPlugins,
            FpsOverlayPlugin {
                config: FpsOverlayConfig {
                    text_config: TextStyle {
                        // Here we define size of our overlay
                        font_size: 50.0,
                        // We can also change color of the overlay
                        color: Color::srgb(0.0, 1.0, 0.0),
                        // If we want, we can use a custom font
                        font: default(),
                    },
                },
            },
            PlayerPlugin,
        ))
        .add_systems(Startup, setup)
        .run();
}

/// set up a simple 3D scene
fn setup(
    mut commands: Commands,
    mut meshes: ResMut<Assets<Mesh>>,
    mut materials: ResMut<Assets<StandardMaterial>>,
) {

    let mut noise = FastNoiseLite::new();
    noise.set_noise_type(Some(NoiseType::OpenSimplex2));
    noise.set_seed(Some(420));

    let cube_mesh = meshes.add(Cuboid::new(1.0, 1.0, 1.0));
    let material = materials.add(StandardMaterial{
        cull_mode: Some(Face::Back),        // 剔除背面
        ..default()
    });

    const CHUNK_WIDTH: usize = 128;
    const CHUNK_HEIGHT: usize = 128;
    let mut noise_data = [[0.;CHUNK_HEIGHT]; CHUNK_WIDTH];

    let mut block_xyz:Vec<Block> = Vec::new();

    for x in 0..CHUNK_WIDTH {
        for z in 0..CHUNK_HEIGHT {
            let negative_1_to_1 = noise.get_noise_2d(x as f32, z as f32);
            noise_data[x][z] = (negative_1_to_1 + 1.) / 2.;

            // 根据noise值生成方块高度
            let y = noise_data[x][z] * 32.;

            //Y轴高度取整，不然方块有间隙
            let y = y.floor() as u32;
            // println!("y的值：{}", y);
            for y in 0..y {
                block_xyz.push(Block{
                    block_id: 1,
                    block_type: BlockType{
                        block_name: "cube".to_string(),
                        x: x as u32,
                        y: y,
                        z: z as u32,
                    },
                });
            }
        }
    }

    for i in block_xyz.iter() {
        commands.spawn(PbrBundle {
            mesh: cube_mesh.clone(),
            material: material.clone(),
            transform: Transform::from_xyz(i.block_type.x as f32, i.block_type.y as f32, i.block_type.z as f32),
            ..default()
        });
    }
}`

**Comment by MrTwoc** [comment]
![cod1](https://github.com/user-attachments/assets/1bc2da20-8920-41f0-a795-9fffbff4ef93)

sorry，this all Code

**Reply by Blatko1** [reply]
Have you tried running the app in the '--release' mode? If not, try: "cargo run --release".

**Reply by MrTwoc** [reply]
Thank you, I haven't been using `--release`. I tried it, and it's still x128*y32*z128, with the FPS being around 24 FPS... It doesn't seem to make much of a difference.
this is my toml

`
[package]
name = "bevygame1"
version = "0.1.0"
edition = "2021"

[dependencies]
bevy = { version = "0.14", features = [
  "dynamic_linking",
  "bevy_dev_tools",
  "shader_format_spirv",
]}
bevy_flycam = "*"
fastnoise-lite = "1.1.1"

[profile.dev]
opt-level = 1
[profile.dev.package."*"]
opt-level = 3
`

**Reply by Blatko1** [reply]
I am unfamiliar with the `Bevy` **renderer**, but I think the problem is when you use the `commands.spawn(PbrBundle { ... })` for each singular block. I think it creates a separate render call for each block, and too many render calls (render passes) are not good. This means that you are creating approximately 262144 (128x16x128) render pass calls, which is definitely not right.
Usually, all those blocks should be stored in a single mesh object, or you should split those blocks into chunks (for example, dimensions 16x16x16, similar to Minecraft chunks).
But still, I am not familiar with the `Bevy` **renderer**, so I am just speculating. I hope this helps!

**Reply by MrTwoc** [reply]
Yes, I feel that using chunks would also be the way to go, but I haven't found any tutorials or demos about using chunks in Bevy. This might help.

**Reply by ray1422** [reply]
Maybe you can try to use the profiling tool to find where the bottlenecks of your systems.

Ref: https://github.com/bevyengine/bevy/blob/main/docs/profiling.md

**Comment by ray1422** [comment]
Also, there are existing Minecraft clone project as your reference 

https://github.com/PhaestusFox/MindCraft

**Reply by MrTwoc** [reply]
Thank you, I'll go check it out right away.

**Comment by Touma-Kazusa2** [comment]
I found a library: https://github.com/splashdust/bevy_voxel_world. It may help you.
Also, here's a video: https://www.youtube.com/watch?v=Ab8TOSFfNp4&list=RDCMUCwe6kcllhainHICL1vnq41g&start_radio=1

I'm feeling a bit lazy, so let me speak Chinese (since there is Chinese in your code).

我最近也在学习关于render的知识,以下是我的想法
帧率低的原因有很多种可能,比如说关于剔除,除了一个方块内部的面需要被剔除,如果两个方块相互靠近,那么它们相接的面也该被剔除,因为这样可以减少draw的开销,除此之外,还有些我不清楚的优化
那么究竟该怎么做呢,说实话我也不太清楚,可能你要自定义一个renderItem类型来实现特殊的渲染(https://github.com/bevyengine/bevy/discussions/9897),
我认为这可能会有些难度,或许你可以看看我开始提到的那个库他是怎么做的,而且重新造轮子意义不大,你直接用他的得了,如果你真要深入的话,可以看看https://jinleili.github.io/learn-wgpu-zh/ ,和bilibili的games104什么的

**Reply by MrTwoc** [reply]
感谢！我回去看看那个games104视频的，
另外那个voxel_world我在逛bevy assets时也看见了，感觉也有些性能问题(地形生成速度赶不上摄像机的速度)，
我找到了一个能看到当前运行参数的工具感觉挺有用的：https://crates.io/crates/bevy-inspector-egui。。
我在B站看到了一个优化方向的视频，我也打算借鉴一下(还是想自己折腾一下)：https://www.bilibili.com/video/BV1HH4y1A7nu?spm_id_from=333.788.videopod.sections&vd_source=511b084e4bf87d71d725c5db0fb20b7f
